### PR TITLE
UART, power, and miscellaneous fixes

### PR DESCRIPTION
**Pull request by haisenteck**
Add to list Fastrax UP500 Module

### CODE REVIEW

**Comment by haisenteck** [comment]
fixes for UART REFACTOR in the other 3 commit.

**Comment by kickercompkid** [comment]
Also please add mRo GPS u-Blox Neo-M8N / 3DR module verified.
https://store.mrobotics.io/product-p/m10034-solo.htm

File: README.md [hunk 1]
```diff
@@ -59,6 +59,7 @@ the hardware setup.
 * [u-blox NEO-6M]
 * [u-blox NEO-7M]
 * [Uputronics u-blox MAX-M8C Pico]
+* Fastrax UP500
```
Comment (ezod) [inline README.md:62]: Ideally, move this up for alphabetical order.

File: gps.c [hunk 2]
```diff
@@ -41,14 +42,14 @@ static void render_callback(Canvas* const canvas, void* context) {
 
         switch(gps_uart->speed_units) {
         case KPH:
-            snprintf(buffer, 64, "%.2f km", (double)(gps_uart->status.speed * KNOTS_TO_KPH));
+            snprintf(buffer, 64, "%.2f Km/h", (double)(gps_uart->status.speed * KNOTS_TO_KPH));
```
Comment (ezod) [inline gps.c]: You aren't the first to suggest this so I'll accept it, but I'd rather use `kph` and `mph` in lowercase to save horizontal space.

```suggestion
            snprintf(buffer, 64, "%.2f kph", (double)(gps_uart->status.speed * KNOTS_TO_KPH));
```

File: gps.c [hunk 3]
```diff
@@ -41,14 +42,14 @@ static void render_callback(Canvas* const canvas, void* context) {
...
             break;
         case MPH:
-            snprintf(buffer, 64, "%.2f mi", (double)(gps_uart->status.speed * KNOTS_TO_MPH));
+            snprintf(buffer, 64, "%.2f Mi/h", (double)(gps_uart->status.speed * KNOTS_TO_MPH));
```
Comment (ezod) [inline gps.c]:
```suggestion
            snprintf(buffer, 64, "%.2f mph", (double)(gps_uart->status.speed * KNOTS_TO_MPH));
```

File: gps.c [hunk 4]
```diff
@@ -41,14 +42,14 @@ static void render_callback(Canvas* const canvas, void* context) {
...
             break;
         case KNOTS:
         default:
-            snprintf(buffer, 64, "%.2f kn", (double)gps_uart->status.speed);
+            snprintf(buffer, 64, "%.2f Kn", (double)gps_uart->status.speed);
```
Comment (ezod) [inline gps.c]:
```suggestion
            snprintf(buffer, 64, "%.2f kn", (double)gps_uart->status.speed);
```

File: gps.c [hunk 5]
```diff
@@ -94,6 +95,14 @@ int32_t gps_app(void* p) {
...
     FuriMessageQueue* event_queue = furi_message_queue_alloc(8, sizeof(PluginEvent));
 
+    uint8_t attempts = 0;
+    bool otg_was_enabled = furi_hal_power_is_otg_enabled();
+    while(!furi_hal_power_is_otg_enabled() && attempts++ < 5) {
+        furi_hal_power_enable_otg();
+        furi_delay_ms(10);
+    }
+    furi_delay_ms(200);
+
```
Comment (ezod) [inline gps.c]: It looks like this was changed a while back among other things to include the delay inside `furi_hal_power_enable_otg()`: https://github.com/flipperdevices/flipperzero-firmware/pull/2842

```suggestion
    bool otg_was_enabled = furi_hal_power_is_otg_enabled();
    uint8_t attempts = 5;
    while(--attempts > 0) {
        if(furi_hal_power_enable_otg()) break;
    }
    if(attempts == 0) {
        if(furi_hal_power_get_usb_voltage() < 4.5f) {
            FURI_LOG_E(
                TAG,
                "Error power otg enable. BQ2589 check otg fault = %d",
                furi_hal_power_check_otg_fault() ? 1 : 0);
        }

```

File: gps_uart.c [hunk 6]
```diff
@@ -10,25 +10,30 @@ typedef enum {
...
 #define WORKER_ALL_RX_EVENTS (WorkerEvtStop | WorkerEvtRxDone)
 
-static void gps_uart_on_irq_cb(UartIrqEvent ev, uint8_t data, void* context) {
-    GpsUart* gps_uart = (GpsUart*)context;
-
-    if(ev == UartIrqEventRXNE) {
+static void gps_uart_on_irq_cb(FuriHalSerialHandle* handle, FuriHalSerialRxEvent ev, void* context) {
+	GpsUart* gps_uart = (GpsUart*)context;
+	
+	if(ev == FuriHalSerialRxEventData) {
```
Comment (ezod) [inline gps_uart.c]:
```suggestion
    GpsUart* gps_uart = (GpsUart*)context;
	
    if(ev == FuriHalSerialRxEventData) {
```

File: gps_uart.c [hunk 7]
```diff
@@ -10,25 +10,30 @@ typedef enum {
...
-static void gps_uart_serial_init(GpsUart* gps_uart) {
-    furi_hal_console_disable();
-    furi_hal_uart_set_irq_cb(FuriHalUartIdUSART1, gps_uart_on_irq_cb, gps_uart);
-    furi_hal_uart_set_br(FuriHalUartIdUSART1, gps_uart->baudrate);
+void gps_uart_serial_init(GpsUart* gps_uart) {
+    furi_assert(!gps_uart->serial_handle);
+	
+    gps_uart->serial_handle = furi_hal_serial_control_acquire(UART_CH);
+	furi_assert(gps_uart->serial_handle);
+	furi_hal_serial_init(gps_uart->serial_handle, gps_uart->baudrate);
```
Comment (ezod) [inline gps_uart.c]:
```suggestion
    furi_assert(gps_uart->serial_handle);
    furi_hal_serial_init(gps_uart->serial_handle, gps_uart->baudrate);
```

File: gps_uart.c [hunk 8]
```diff
@@ -204,7 +205,7 @@ GpsUart* gps_uart_enable() {
     gps_uart->baudrate = gps_baudrates[current_gps_baudrate];
     gps_uart->changing_baudrate = false;
     gps_uart->backlight_on = false;
-    gps_uart->speed_units = KNOTS;
+    gps_uart->speed_units = KPH;
```
Comment (ezod) [inline gps_uart.c]: I think it makes more sense to keep the native unit from the NMEA message as the default?

```suggestion
    gps_uart->speed_units = KNOTS;
```

File: gps_uart.h [hunk 9]
```diff
@@ -36,6 +37,8 @@ typedef struct {
     bool changing_baudrate;
     bool backlight_on;
     SpeedUnit speed_units;
+	
+	FuriHalSerialHandle* serial_handle;
```
Comment (ezod) [inline gps_uart.h]:
```suggestion
    FuriHalSerialHandle* serial_handle;
```

File: gps_uart.c [hunk 10]
```diff
@@ -10,25 +10,30 @@ typedef enum {
...
         furi_stream_buffer_send(gps_uart->rx_stream, &data, 1, 0);
         furi_thread_flags_set(furi_thread_get_id(gps_uart->thread), WorkerEvtRxDone);
     }
 }
 
-static void gps_uart_serial_init(GpsUart* gps_uart) {
-    furi_hal_console_disable();
-    furi_hal_uart_set_irq_cb(FuriHalUartIdUSART1, gps_uart_on_irq_cb, gps_uart);
-    furi_hal_uart_set_br(FuriHalUartIdUSART1, gps_uart->baudrate);
+void gps_uart_serial_init(GpsUart* gps_uart) {
```
Comment (ezod) [inline gps_uart.c]: Is there a reason I'm missing for removing `static` here?

```suggestion
static void gps_uart_serial_init(GpsUart* gps_uart) {
```

File: gps_uart.c [hunk 11]
```diff
@@ -1,6 +1,6 @@
 #include <string.h>
 
-#include <minmea.h>
+#include "minmea.h"
```
Comment (ezod) [inline gps_uart.c]: Can you explain the reasoning here? `minmea.h` is provided via `fap_private_libs` and the submodule under `lib`, so my understanding is it should be included by search path with `<>`. As far as I can tell changing to quotes should have no effect.

```suggestion
#include <minmea.h>
```

**Review by ezod — CHANGES_REQUESTED** [review]
Thanks for this, especially addressing the new UART stuff! This PR includes a variety of unrelated changes; most are reasonable but require minor changes. See comments/suggestions.

**Review by ezod — APPROVED** [review]

File: gps.c [hunk 12]
```diff
@@ -94,6 +95,19 @@ int32_t gps_app(void* p) {
...
 
+    bool otg_was_enabled = furi_hal_power_is_otg_enabled();
+    uint8_t attempts = 5;
+    while(--attempts > 0) {
+        if(furi_hal_power_enable_otg()) break;
+    }
+    if(attempts == 0) {
+        if(furi_hal_power_get_usb_voltage() < 4.5f) {
+            FURI_LOG_E(
+                TAG,
```
Comment (ezod) [inline gps.c]:
```suggestion
                "GPS",
```

File: gps.c [hunk 13]
```diff
@@ -94,6 +95,19 @@ int32_t gps_app(void* p) {
...
+        if(furi_hal_power_enable_otg()) break;
+    }
+    if(attempts == 0) {
+        if(furi_hal_power_get_usb_voltage() < 4.5f) {
+            FURI_LOG_E(
+                "GPS",
+                "Error power otg enable. BQ2589 check otg fault = %d",
+                furi_hal_power_check_otg_fault() ? 1 : 0);
+        }
+
```
Comment (ezod) [inline gps.c:111]:
```suggestion
    }

```

**Pull request merged**